wpilibj/src/main/java/edu/wpi/first/wpilibj/controller/PIDController.java

```diff
@@ -0,0 +1,577 @@
+  public void setPercentTolerance(double tolerance) {
+    setPercentTolerance(tolerance, Double.POSITIVE_INFINITY);
+  }
+
+  /**
+   * Sets the percent error which is considered tolerable for use with atSetpoint().
+   *
+   * @param tolerance      Percent error which is tolerable.
+   * @param deltaTolerance Change in percent error per second which is tolerable.
+   */
+  public void setPercentTolerance(double tolerance, double deltaTolerance) {
+    m_thisMutex.lock();
+    try {
+      m_toleranceType = Tolerance.kPercent;
+      m_tolerance = tolerance;
+      m_deltaTolerance = deltaTolerance;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the difference between the setpoint and the measurement.
+   *
+   * @return The error.
+   */
+  public double getError() {
+    m_thisMutex.lock();
+    try {
+      return getContinuousError(m_currError);
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the change in error per second.
+   */
+  public double getDeltaError() {
+    double error = getError();
+
+    m_thisMutex.lock();
+    try {
+      return (error - m_prevError) / getPeriod();
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Calculates the output of the PID controller.
+   *
+   * @param measurement The current measurement of the process variable.
+   * @return The controller output.
+   */
+  @SuppressWarnings("LocalVariableName")
+  public double calculate(double measurement) {
+    // Storage for function inputs
+    final double Kp;
+    final double Ki;
+    final double Kd;
+    final double minimumOutput;
+    final double maximumOutput;
+
+    // Storage for function input-outputs
+    double totalError;
+
+    m_thisMutex.lock();
+    try {
+      Kp = m_Kp;
+      Ki = m_Ki;
+      Kd = m_Kd;
+      minimumOutput = m_minimumOutput;
+      maximumOutput = m_maximumOutput;
+
+      m_prevError = m_currError;
+      m_currError = getContinuousError(m_setpoint - measurement);
+      totalError = m_totalError;
+    } finally {
+      m_thisMutex.unlock();
+    }
+
+    if (Ki != 0) {
+      totalError = clamp(totalError + m_currError * getPeriod(), minimumOutput / Ki,
+          maximumOutput / Ki);
+    }
+
+    double output =
+        clamp(Kp * m_currError + Ki * totalError + Kd * (m_currError - m_prevError) / getPeriod(),
+            minimumOutput, maximumOutput);
+
+    m_thisMutex.lock();
+    try {
+      m_totalError = totalError;
+      m_output = output;
+    } finally {
+      m_thisMutex.unlock();
+    }
+
+    return output;
+  }
+
+  /**
+   * Reads the input, calculate the output accordingly, and return the output.
+   *
+   * @param measurement The current measurement of the process variable.
+   * @param setpoint The setpoint of the controller.
+   * @return The controller output.
+   */
+  public double calculate(double measurement, double setpoint) {
+    setSetpoint(setpoint);
+    return calculate(measurement);
+  }
```

Comment:
There is another race condition here because the setpoint can change between the call to setSetpoint and calculate causing this method to break the contract.